snap-core/src/main/java/org/esa/snap/core/dataio/dimap/VariableCache.java

```diff
@@ -0,0 +1,94 @@
+package org.esa.snap.core.dataio.dimap;
+
+import org.esa.snap.core.datamodel.Band;
+import org.esa.snap.core.datamodel.ProductData;
+
+import javax.imageio.stream.ImageOutputStream;
+import java.awt.Dimension;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+class VariableCache {
+
+    private static int LINES_PER_BUFFER = 128;
+    // package access for testing only in this first version - refactor later tb 2020-06-29
+    final CacheBlock[] cacheBlocks;
+    private final List<Integer> completedIndices;
+
+    private final int width;
+    private final int rasterHeight;
+    private final int dataType;
+    private final double noDataValue;
+
+    VariableCache(Band variable) {
+        final Dimension rasterSize = variable.getRasterSize();
+        width = rasterSize.width;
+        rasterHeight = rasterSize.height;
+        dataType = variable.getDataType();
+        noDataValue = variable.getNoDataValue();
+
+        final int numCacheBlocks = (rasterSize.height - 1) / LINES_PER_BUFFER + 1;
+        cacheBlocks = new CacheBlock[numCacheBlocks];
+        completedIndices = new ArrayList<>();
+    }
+
+    boolean update(int sourceOffsetX, int sourceOffsetY, int sourceWidth, int sourceHeight, ProductData sourceBuffer) {
+        final int firstIdx = sourceOffsetY / LINES_PER_BUFFER;
+        final int lastIdx = (sourceOffsetY + sourceHeight - 1) / LINES_PER_BUFFER;
+
+        boolean canWrite = false;
+        int yReadOff = 0;
+        for (int index = firstIdx; index <= lastIdx; index++) {
+            final int remainingHeight = sourceHeight - yReadOff;
+            final int blockHeight = remainingHeight > LINES_PER_BUFFER ? LINES_PER_BUFFER : remainingHeight;
```

Comment:
line 44 could be replaced by:
`final int blockHeight = Math.min(remainingHeight, LINES_PER_BUFFER);`

Reply:
right